src/ring_theory/tensor_product.lean

```diff
@@ -822,3 +822,63 @@ begin
   rw [←E1.range_val, ←E2.range_val, ←algebra.tensor_product.product_map_range],
   exact (algebra.tensor_product.product_map E1.val E2.val).to_linear_map.finite_dimensional_range,
 end
+
+namespace tensor_product.algebra
+
+variables {R A B M : Type*}
+variables [comm_semiring R] [add_comm_monoid M] [module R M]
+variables [semiring A] [semiring B] [module A M] [module B M]
+variables [algebra R A] [algebra R B]
+variables [is_scalar_tower R A M] [is_scalar_tower R B M]
+
+/-- An auxiliary definition, used for constructing the `module (A ⊗[R] B) M` instance below. -/
+def module_aux : A ⊗[R] B →ₗ[R] M →ₗ[R] M :=
+tensor_product.lift
+{ to_fun := λ a, a • (algebra.lsmul R M : B →ₐ[R] module.End R M).to_linear_map,
+  map_add' := λ r t, by { ext, simp only [add_smul, linear_map.add_apply] },
+  map_smul' := λ n r, by { ext, simp only [ring_hom.id_apply, linear_map.smul_apply, smul_assoc] } }
+
+lemma module_aux_apply (a : A) (b : B) (m : M) :
+  module_aux (a ⊗ₜ[R] b) m = a • b • m :=
+by simp [module_aux]
+
+variables [smul_comm_class A B M]
+
+/-- If `M` is a representation of two different `R`-algebras `A` and `B` whose actions commute,
+then it is a representation the `R`-algebra `A ⊗[R] B`.
+
+A important example arises from a semiring `S`; allowing `S` to act on itself via left and right
+multiplication, the roles of `R`, `A`, `B`, `M` are played by `ℕ`, `S`, `Sᵐᵒᵖ`, `S`. -/
+instance : module (A ⊗[R] B) M :=
+{ smul := λ x m, module_aux x m,
```

Comment:
Reiterating what I asked on zulip; does this create a diamond with `semiring.to_module` when `M = A ⊗[R] B`?

Reply:
Our comments just crossed, I was replying [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/ideals.20in.20non-comm.20rings/near/284123987) as you wrote this

Reply:
Oooh wait, I think I was considering a different diamond. I think we _do_ get a diamond here. Let me just check.

Reply:
Wow, great point. Indeed it comes down to the following:
```lean
example {R A} [comm_semiring R] [comm_semiring A] [algebra R A]
  (a b : A) : a ⊗ₜ[R] b = 1 ⊗ₜ[R] (a * b) :=
sorry -- This is false!
```
The term on the LHS has nothing whatsoever to do with the product on `A` so this cannot be true.

For a counterexample I guess one can take `R` to be a field and `A` to be a commutative `R`-algebra with two linearly-independent elements that multiply together to give zero.

Reply:
I see three possible ways forward here:

1. Live with this diamond
2. Make `tensor_product.algebra.module` a `def` and a local `instance` in the new file `bimodule.lean`
3. Carry out the work I outlined ages ago [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.234773.20base.20change/near/240929258) which resolves the fundamental ambiguity in lifting a commuting scalar action carried by `M` or `N` to `M ⊗[R] N`

I believe 3 will ultimately be necessary but I think both 1 and especially 2 are viable now.

Thoughts?

Reply:
Some initial thoughts: I think the more specialized instance of the form `module (A ⊗[R] Bᵐᵒᵖ) M` would probably be safe enough, because in the long run we probably want the right actions on `A ⊗[R] B` to act on the B part (making right actions on A inaccessible).

Reply:
I opted to handle the diamond conservatively by making the `instance` a `def`.